llvm/lib/Passes/PassBuilder.cpp

```diff
@@ -528,6 +529,106 @@ PassBuilder::PassBuilder(TargetMachine *TM, PipelineTuningOptions PTO,
 #include "llvm/Passes/MachinePassRegistry.def"
     });
   }
+  auto parseLevelParam = [](StringRef P) -> Expected<OptimizationLevel> {
+    auto OptLevel = llvm::StringSwitch<std::optional<OptimizationLevel>>(P)
+                        .Case("O0", OptimizationLevel::O0)
+                        .Case("O1", OptimizationLevel::O1)
+                        .Case("O2", OptimizationLevel::O2)
+                        .Case("O3", OptimizationLevel::O3)
+                        .Case("Os", OptimizationLevel::Os)
+                        .Case("Oz", OptimizationLevel::Oz)
+                        .Default(std::nullopt);
+    if (!OptLevel)
+      return llvm::createStringError(llvm::inconvertibleErrorCode(),
+                                     "invalid optimization level '%s'",
+                                     P.str().c_str());
+    return *OptLevel;
+  };
+
+  // Module-level callbacks without LTO phase
+  registerPipelineParsingCallback(
+      [this, parseLevelParam](StringRef Name, ModulePassManager &PM,
+                              ArrayRef<PassBuilder::PipelineElement>) {
+#define MODULE_CALLBACK(NAME, INVOKE)                                          \
+  if (PassBuilder::checkParametrizedPassName(Name, NAME)) {                    \
+    auto L = PassBuilder::parsePassParameters(parseLevelParam, Name, NAME);    \
+    if (!L)                                                                    \
+      return (errs() << NAME ": " << toString(L.takeError()) << '\n', false);  \
```

Comment:
could we use a more conventional
```
if (!L) {
  errs() << ...
  return false;
}
```
this macro already has multiple lines, I don't think it matters if we add another one.